cmd/juju/cloud/add.go

```diff
@@ -90,27 +107,177 @@ func (c *addCloudCommand) Run(ctxt *cmd.Context) error {
 	if !ok {
 		return errors.Errorf("cloud %q not found in file %q", c.Cloud, c.CloudFile)
 	}
-	publicClouds, _, err := cloud.PublicCloudMetadata()
+
+	if err := c.verifyName(c.Cloud); err != nil {
+		return errors.Trace(err)
+	}
+
+	return addCloud(c.Cloud, newCloud)
+}
+
+func (c *addCloudCommand) runInteractive(ctxt *cmd.Context) error {
+	allproviders := environs.RegisteredProviders()
+
+	public, _, err := cloud.PublicCloudMetadata()
 	if err != nil {
 		return err
 	}
-	if _, ok = publicClouds[c.Cloud]; ok && !c.Replace {
-		return errors.Errorf("%q is the name of a public cloud; use --replace to use your cloud definition instead", c.Cloud)
+	personal, err := cloud.PersonalCloudMetadata()
+	if err != nil {
+		return err
+	}
+
+	var unsupported []string
+	var providers []string
+	for _, name := range allproviders {
+		provider, err := environs.Provider(name)
+		if err != nil {
+			// should be impossible
+			return errors.Trace(err)
+		}
+
+		if provider.CloudSchema() != nil {
+			providers = append(providers, name)
+		} else {
+			unsupported = append(unsupported, name)
+		}
+	}
+	sort.Strings(providers)
+
+	supportedCloud := interact.VerifyOptions("cloud type", providers, false)
+
+	cloudVerify := func(s string) (ok bool, errmsg string, err error) {
+		ok, errmsg, err = supportedCloud(s)
+		if err != nil {
+			return false, "", errors.Trace(err)
+		}
+		if ok {
+			return true, "", nil
+		}
+		// Print out a different message if they entered a valid provider that
+		// just isn't something we want people to add (like ec2).
+		for _, name := range unsupported {
+			if strings.ToLower(name) == strings.ToLower(s) {
+				return false, fmt.Sprintf("Cloud type %q not supported for interactive add-cloud.", s), nil
+			}
+		}
+		return false, errmsg, nil
+	}
+
+	w := ansiterm.NewWriter(ctxt.Stdout)
+	pollster := interact.New(ctxt.Stdin, ctxt.Stdout, errWriter{w})
+	cloudType, err := pollster.SelectVerify(interact.List{
+		Singular: "cloud type",
+		Plural:   "cloud types",
+		Options:  providers,
+	}, cloudVerify)
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var name string
+	for {
+		name, err = pollster.Enter("a name for the cloud")
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = nameExists(name, public, personal)
+		if err == nil {
+			break
+		}
+		// we do this instead of returning the error message
+		override, err2 := pollster.YN(err.Error()+", do you want to override that definition", false)
```

Comment:
it's destructive if it's overriding one you created, but not destructive if it's overriding a default one (like EC2).  Although, in the interactive case, we don't even support overriding a default one, so that point is kind of moot.  So, yes, it's destructive.  I can make the error message a little scarier.